### ISSUE TITLE
Change bundling process for jszip

### ISSUE THREAD

**Issue by Beak-man** [NASAWorldWind/WebWorldWind#202]
The third party library jszip, used for .kmz support in WebWW, has Browserify as bundling tool. When the library is built, Browserify links every module with some sort of nested CommonJS-like interfaces inside the file. [Webpack and Node seem to have issues with this linking method](https://github.com/webpack/webpack/issues/319). When Web WorldWind is consumed as an npm package, the problem with this kind of interfaces arise.

The same problem was happening with another 3rd party library, proj4-src. The authors of this latter library moved away from Browserify as bundling tool, so the issue with WebWW's npm package was avoided by updating the library and changing is interface to a pure AMD one, in order to get it in line with every other module of WebWW. But jszip is still being built with Browserify in its last version.

Changing the bundling method for jszip will be considered and investigated in order to bundle the library in a different manner (and hopefully, getting an AMD module as output). Webpack seems to be the ideal solution, SystemJS will be considered.

**Comment by Beak-man** [comment]
This task will be limited for one day and a half to verify feasibility. It may also leads us into valuable learning on migrating bundling tools for WebWW itself in the future.

**Comment by Beak-man** [comment]
A pure AMD version of jszip was successfully created by changing jszip bundling tool from Browserify to Webpack.

I forked jszip. It has to be noted that everything that follows was done in the 'v2.x' branch of jszip:

https://github.com/Beak-man/jszip/tree/v2.x

This due to the fact that the latest 3.x version of jszip in the master branch has some breaking changes in relation to the version that was used in WebWW for .kmz file support. I did test v3.x and tried to address these changes:

https://stuk.github.io/jszip/documentation/upgrade_guide.html

But that would have taken some time, and that's outside the scope of this task. So I stuck with using v2.x. Before getting the AMD version with webpack, I produced a library with its current Browserify bundler with the following command:

`browserify ./lib/index.js > ./dist/jszip.js --standalone JSZip`

The '--standalone' option produces the UMD interface with an option for a global object in case it's directly loaded in the browser with a script tag, as noted here:

https://github.com/substack/node-browserify#usage

Without this option, [line #82 of /src/formats/kml/KmlFile.js in WebWW](https://github.com/NASAWorldWind/WebWorldWind/blob/develop/src/formats/kml/KmlFile.js#L82) wasn't working:

````javascript
var kmzFile = new JsZip();
````

I'm suspicious as to why this happened. It suggests that jszip is currently being loaded as a global object. After using the '--standalone' modifier, I dropped the result in Web WorldWind's 'util' folder and it worked, but of course this version still has the the UMD and nested interfaces that's giving us problems with npm packing. 

So I worked into producing a pure AMD version of the library.

WebPack is able to both consume and output in many types of JS modules, and it can output a pure AMD file: https://webpack.github.io/docs/configuration.html#output-librarytarget

I added WebPack as a dev dependency and included the following configuration for WebPack in the file `webpack.conf.js`. The 'libraryTarget' attribute is, of course, what instructs WebPack to produce an AMD module as output:

````javascript
module.exports = {
    entry: "./lib/index.js",

    output: {
        filename: "./dist/jszip.js",
        library: "JSZip",
        libraryTarget: "amd"
    }
}
````
After downloading my fork, the command `npm install` followed by `webpack` should generate the AMD version of jszip in the /dist folder. It's already present in the repository:

https://github.com/Beak-man/jszip/blob/v2.x/dist/jszip.js

This file is currently being tested. Again, I'm getting an error with the `JsZip()` constructor in KmlFile.js, but I expect to address this in the following hours.

**Comment by pdavidc** [comment]
Thanks @Beak-man, this is an encouraging update. It seems worth continuing on this task through today, then re-evaluating depending on what you find.

**Comment by Beak-man** [comment]
The library is working now. One Webpack configuration parameter introduced an argument to the library that wasn't part of the RequireJS specification.

Specifically speaking, this `webpack.conf.js` file:
````javascript
module.exports = {
    entry: "./lib/index.js",

    output: {
        filename: "./dist/jszip.js",
        library: "JSZip",
        libraryTarget: "amd"
    }
}
````
Produced a library with the following arguments:
````javascript
define("JSZip", [], function() { return /******/ (function(modules) { ... }
````
According to [Webpack's configuration docs](https://webpack.github.io/docs/configuration.html#output-library), the 'library' attribute should be used when creating libraries, but in this case it introduced that "JSZip" argument that is not recognized by our RequireJS bundler. By removing the 'library' attribute from the configuration file:
````javascript
module.exports = {
    entry: "./lib/index.js",

    output: {
        filename: "./dist/jszip.js",
        libraryTarget: "amd"
    }
}
````
 We obtain the library with the usual 'define' arguments:
````javascript
define(b[], function() { return /******/ (function(modules) { ... }
````
Now the npm loads, even with the shim.

**Comment by Beak-man** [comment]
.kml support works as expected in [the React tester](https://github.com/Beak-man/worldwind-react/tree/feature/kml_test), (kml_test branch) with the latest npm package. Zipped files are not being tested as of now, but the KmlFile object (which includes the jszip library) loads correctly now:

![kmlreact](https://user-images.githubusercontent.com/7622894/29673649-7c3b8276-88b5-11e7-82a7-a4c0473584f3.png)

**Comment by Beak-man** [comment]
This following is a guide on how to obtain the AMD version of jszip. It can be obtained either from the author's GitHub or from my fork of it.

[From the author's GitHub](https://github.com/Stuk/jszip):

First we need to install Webpack globally:

`> npm install webpack -g`

Then we clone the repo and switch to the v2.x branch. The latest v3.x version doesn't work with our current examples since it introduced some breaking changes, [as documented here](https://stuk.github.io/jszip/documentation/upgrade_guide.html):

`> git fetch`
`> git checkout v2.x`

We can see that the bundled library is present in the /dist folder (the jszip.js file), but this is the version bundled with Browserify. We will indicate webpack to overwrite it with an AMD version of it:

`> webpack --entry './lib/index.js' --output-filename './dist/jszip.js'
--output-library-target 'amd'`

This is a little bit verbose and can be simplified, but the extra parameters make it self-explanatory. The index.js file is the entry point of the jszip library, we set jszip.js as output, and we're telling Webpack that we desire an AMD module as library target.

[From my fork of jszip:](https://github.com/Beak-man/jszip)

A file with configuration settings for Webpack can be added to simplify the library generation, which is the `wepack.conf.js` file described in the prior comments of this issue. Such file has been added to my fork of jszip, as well as Webpack itself as a devDependency in package.json.

After cloning it and switching to the v2.x branch as mentioned above, we can install every development dependency (including Webpack):

`npm install`

To get every development dependency (including Webpack), and only execute:

`> webpack`

... to obtain the AMD version of jszip.js in the dist folder. Such file is already present in my fork:
